core/src/main/java/tc/oc/pgm/map/MapInfoImpl.java

```diff
@@ -349,44 +356,64 @@ protected void setContext(MapContextImpl context) {
       // last.
       if (this.gamemodes.isEmpty()) {
-        this.gamemodes =
-            this.tags.stream()
-                .filter(MapTag::isGamemode)
-                .sorted(
-                    Comparator.comparing(MapTag::isAuxiliary)
-                        .thenComparing(Comparator.naturalOrder()))
-                .map(MapTag::getGamemode)
-                .collect(StreamUtils.toImmutableList());
+        this.gamemodes = this.tags.stream()
+            .filter(MapTag::isGamemode)
+            .sorted(
+                Comparator.comparing(MapTag::isAuxiliary).thenComparing(Comparator.naturalOrder()))
+            .map(MapTag::getGamemode)
+            .collect(StreamUtils.toImmutableList());
       }
     }
     this.context = new SoftReference<>(context);
   }
 
-  private static class VariantData implements VariantInfo {
+  private class VariantData implements VariantInfo {
+    // taken from https://minecraft.wiki/w/Data_version#Java_Edition
+    private static final int MAP_DATA_VERSION_1_13 = 1519;
+    private static final Version VERSION_1_13 = new Version(1, 13, 0);
     private final String variantId;
     private final String mapName;
     private final String mapId;
     private final String world;
+    private final Range<Version> serverVersions;
 
     public VariantData(Element root, @Nullable Element variantEl) throws InvalidXMLException {
       String name = assertNotNull(Node.fromRequiredChildOrAttr(root, "name").getValueNormalize());
       String slug = assertNotNull(root).getChildTextNormalize("slug");
+      Node minVer = Node.fromAttr(root, "min-server-version");
+      Node maxVer = Node.fromAttr(root, "max-server-version");
 
       if (variantEl == null) {
         this.variantId = DEFAULT_VARIANT;
         this.mapName = name;
         this.world = null;
       } else {
-        this.variantId = Node.fromRequiredAttr(variantEl, "id").getValue();
-        if (DEFAULT_VARIANT.equals(variantId)) {
-          throw new InvalidXMLException("Default variant is not allowed", variantEl);
-        }
+        this.variantId = XMLUtils.parseRequiredId(variantEl);
+        if (DEFAULT_VARIANT.equals(variantId))
+          throw new InvalidXMLException("Variant id must not be 'default'", variantEl);
+
         boolean override = XMLUtils.parseBoolean(Node.fromAttr(variantEl, "override"), false);
         this.mapName = (override ? "" : name + ": ") + variantEl.getTextNormalize();
         this.world = variantEl.getAttributeValue("world");
-        if (slug != null) slug += "_" + variantId;
+
+        String variantSlug = variantEl.getAttributeValue("slug");
+        if (variantSlug != null) slug = variantSlug;
+        else if (override) slug = null;
+        else if (slug != null) slug += "_" + variantId;
+
+        minVer = fallback(Node.fromAttr(variantEl, "min-server-version"), minVer);
+        maxVer = fallback(Node.fromAttr(variantEl, "max-server-version"), maxVer);
```

Comment:
supposed to be other way around? (otherwise root-level map settings will take priority over variant-level settings)
```suggestion
        minVer = fallback(minVer, Node.fromAttr(variantEl, "min-server-version"));
        maxVer = fallback(maxVer, Node.fromAttr(variantEl, "max-server-version"));
```

Reply:
Nope, it takes the first non null param

Reply:
Currently, if both the default variant and a variant define a version bound (such as the example in the cover letter), the first non-null parameter in that case the root-level `{min,max}-server-version`, i. e. the `{min,max}-server-version` in `<map>` would take precedence over the attributes in `<variant>`. 

In other words, root-level/default variant version bound is used first, falling back to the variant-level bound if not present 🤷‍♀️

Reply:
This was observed and read incorrectly on my side, sorry! (╯°□°)╯︵ ┻━┻


There's actually a different issue with this – the root-level version bounds get unused node warnings when a variant overrides them… although that's correct, that should get silenced as it is intended to not be used